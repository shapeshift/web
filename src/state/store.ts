import { configureStore } from '@reduxjs/toolkit'
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux'
import { registerSelectors } from 'reselect-tools'

import { reducer, ReduxState } from './reducer'
import { assetApi } from './slices/assetsSlice/assetsSlice'
import { marketApi } from './slices/marketDataSlice/marketDataSlice'
import { portfolioApi } from './slices/portfolioSlice/portfolioSlice'
import * as portfolioSelectors from './slices/portfolioSlice/selectors'

registerSelectors(portfolioSelectors)

const apiMiddleware = [portfolioApi.middleware, marketApi.middleware, assetApi.middleware]

/// This allows us to create an empty store for tests
export const createStore = () =>
  configureStore({
    reducer,
    middleware: getDefaultMiddleware =>
      getDefaultMiddleware({
        immutableCheck: { warnAfter: 128 },
        serializableCheck: { warnAfter: 128 }
      }).concat(apiMiddleware),
    devTools: true
  })

export const store = createStore()

export const useAppSelector: TypedUseSelectorHook<ReduxState> = useSelector

// https://redux-toolkit.js.org/usage/usage-with-typescript#getting-the-dispatch-type
export type AppDispatch = typeof store.dispatch
export const useAppDispatch = () => useDispatch<AppDispatch>()
