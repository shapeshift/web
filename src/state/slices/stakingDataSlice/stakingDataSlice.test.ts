import {
  emptyMockStakingData,
  mockStakingData,
  mockStakingDataWithOnlyRewards,
  mockStakingDataWithOnlyUndelegations,
  mockStakingWithUnknownValidators,
  mockValidatorData,
} from 'test/mocks/stakingData'
import { clearState, store } from 'state/store'

import {
  selectActiveStakingOpportunityDataByAssetId,
  selectNonloadedValidators,
  selectRewardsAmountByAssetId,
  selectSingleValidator,
  selectStakingDataIsLoaded,
  selectTotalBondingsBalanceByAssetId,
  selectTotalStakingDelegationCryptoByAccountSpecifier,
  selectValidatorIsLoaded,
} from './selectors'
import { stakingData } from './stakingDataSlice'

const cosmosAccountSpecifier: string =
  'cosmos:cosmoshub-4:cosmos1wc4rv7dv8lafv38s50pfp5qsgv7eknetyml669'
const otherCosmosAccountSpecifier: string =
  'cosmos:cosmoshub-4:cosmos1j26n3mjpwx4f7zz65tzq3mygcr74wp7kcwcner'
const cosmosAssetId = 'cosmos:cosmoshub-4/slip44:118'
const SHAPESHIFT_VALIDATOR_ADDRESS = 'cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf'

describe('stakingDataSlice', () => {
  const consoleInfoSpy = jest.spyOn(console, 'info').mockImplementation(() => void 0)
  beforeEach(() => {
    clearState()
  })
  afterAll(() => consoleInfoSpy.mockRestore())

  it('returns uninitialized properties for initialState', async () => {
    expect(store.getState().stakingData).toEqual({
      byAccountSpecifier: {},
      byValidator: {},
      status: 'idle',
      validatorStatus: 'idle',
    })
  })

  describe('reducers', () => {
    describe('setStatus', () => {
      it('updates status in state', async () => {
        store.dispatch(stakingData.actions.setStatus('loading'))
        expect(store.getState().stakingData.status).toEqual('loading')
        expect(store.getState().stakingData.validatorStatus).toEqual('idle')

        store.dispatch(stakingData.actions.setStatus('loaded'))
        expect(store.getState().stakingData.status).toEqual('loaded')
        expect(store.getState().stakingData.validatorStatus).toEqual('idle')

        store.dispatch(stakingData.actions.setStatus('idle'))
        expect(store.getState().stakingData.status).toEqual('idle')
        expect(store.getState().stakingData.validatorStatus).toEqual('idle')
      })
    })

    describe('setValidatorStatus', () => {
      it('updates validator status in state', async () => {
        store.dispatch(stakingData.actions.setValidatorStatus('loading'))
        expect(store.getState().stakingData.status).toEqual('idle')
        expect(store.getState().stakingData.validatorStatus).toEqual('loading')

        store.dispatch(stakingData.actions.setValidatorStatus('loaded'))
        expect(store.getState().stakingData.status).toEqual('idle')
        expect(store.getState().stakingData.validatorStatus).toEqual('loaded')

        store.dispatch(stakingData.actions.setValidatorStatus('idle'))
        expect(store.getState().stakingData.status).toEqual('idle')
        expect(store.getState().stakingData.validatorStatus).toEqual('idle')
      })
    })

    describe('upsertStakingData', () => {
      it('updates or inserts staking data in state', async () => {
        store.dispatch(
          stakingData.actions.upsertStakingData({
            accountSpecifier: cosmosAccountSpecifier,
            stakingData: mockStakingData,
          }),
        )
        expect(
          store.getState().stakingData.byAccountSpecifier[cosmosAccountSpecifier],
        ).toMatchSnapshot()
        expect(
          store.getState().stakingData.byAccountSpecifier[otherCosmosAccountSpecifier],
        ).toBeUndefined()
      })

      it('updates or inserts empty staking data in state', async () => {
        store.dispatch(
          stakingData.actions.upsertStakingData({
            accountSpecifier: otherCosmosAccountSpecifier,
            stakingData: emptyMockStakingData,
          }),
        )
        expect(
          store.getState().stakingData.byAccountSpecifier[otherCosmosAccountSpecifier],
        ).toMatchSnapshot()
        expect(
          store.getState().stakingData.byAccountSpecifier[cosmosAccountSpecifier],
        ).toBeUndefined()
      })
    })

    describe('upsertValidatorData', () => {
      it('updates or inserts validator data in state', async () => {
        store.dispatch(
          stakingData.actions.upsertValidatorData({
            validators: mockValidatorData,
          }),
        )
        expect(store.getState().stakingData.byValidator).toMatchSnapshot()
      })
    })

    describe('clear', () => {
      it('clears state back to initial state', async () => {
        store.dispatch(stakingData.actions.setStatus('loaded'))
        store.dispatch(stakingData.actions.setValidatorStatus('loaded'))
        store.dispatch(
          stakingData.actions.upsertStakingData({
            accountSpecifier: cosmosAccountSpecifier,
            stakingData: mockStakingData,
          }),
        )
        store.dispatch(
          stakingData.actions.upsertValidatorData({
            validators: mockValidatorData,
          }),
        )

        expect(
          store.getState().stakingData.byAccountSpecifier[cosmosAccountSpecifier],
        ).toMatchSnapshot()
        expect(store.getState().stakingData.byValidator).toMatchSnapshot()
        store.dispatch(stakingData.actions.clear())

        expect(store.getState().stakingData).toEqual({
          byAccountSpecifier: {},
          byValidator: {},
          status: 'idle',
          validatorStatus: 'idle',
        })
      })
    })
  })

  describe('selectors', () => {
    describe('selectActiveStakingOpportunityDataByAssetId', () => {
      it('returns empty array on initial state', async () => {
        const selected = selectActiveStakingOpportunityDataByAssetId(
          store.getState(),
          cosmosAccountSpecifier,
          SHAPESHIFT_VALIDATOR_ADDRESS,
          cosmosAssetId,
        )
        expect(selected).toEqual([])
      })

      it('returns empty array when staking data is loaded but validators are not', async () => {
        store.dispatch(
          stakingData.actions.upsertStakingData({
            accountSpecifier: cosmosAccountSpecifier,
            stakingData: mockStakingData,
          }),
        )

        const selected = selectActiveStakingOpportunityDataByAssetId(
          store.getState(),
          cosmosAccountSpecifier,
          SHAPESHIFT_VALIDATOR_ADDRESS,
          cosmosAssetId,
        )
        expect(selected).toEqual([])
      })

      describe('validators and staking data are loaded', () => {
        beforeEach(() => {
          store.dispatch(
            stakingData.actions.upsertValidatorData({
              validators: mockValidatorData,
            }),
          )
        })

        it('returns correct array with delegations, undelegations, redelegations and rewards', async () => {
          store.dispatch(
            stakingData.actions.upsertStakingData({
              accountSpecifier: cosmosAccountSpecifier,
              stakingData: mockStakingData,
            }),
          )

          const selected = selectActiveStakingOpportunityDataByAssetId(
            store.getState(),
            cosmosAccountSpecifier,
            SHAPESHIFT_VALIDATOR_ADDRESS,
            cosmosAssetId,
          )
          expect(selected).toMatchSnapshot()
        })

        it('returns correct array with only undelegations', async () => {
          store.dispatch(
            stakingData.actions.upsertStakingData({
              accountSpecifier: cosmosAccountSpecifier,
              stakingData: mockStakingDataWithOnlyUndelegations,
            }),
          )

          const selected = selectActiveStakingOpportunityDataByAssetId(
            store.getState(),
            cosmosAccountSpecifier,
            SHAPESHIFT_VALIDATOR_ADDRESS,
            cosmosAssetId,
          )
          expect(selected).toMatchSnapshot()
        })

        it('returns correct array with only rewards', async () => {
          store.dispatch(
            stakingData.actions.upsertStakingData({
              accountSpecifier: cosmosAccountSpecifier,
              stakingData: mockStakingDataWithOnlyRewards,
            }),
          )

          const selected = selectActiveStakingOpportunityDataByAssetId(
            store.getState(),
            cosmosAccountSpecifier,
            SHAPESHIFT_VALIDATOR_ADDRESS,
            cosmosAssetId,
          )
          expect(selected).toMatchSnapshot()
        })
      })
    })

    describe('selectNonloadedValidators', () => {
      it('returns empty array on initial state', async () => {
        const selected = selectNonloadedValidators(store.getState(), cosmosAccountSpecifier)
        expect(selected).toEqual([])
      })

      it('returns all validators in staking data when staking data is loaded but validators are not', async () => {
        store.dispatch(
          stakingData.actions.upsertStakingData({
            accountSpecifier: cosmosAccountSpecifier,
            stakingData: mockStakingData,
          }),
        )

        const selected = selectNonloadedValidators(store.getState(), cosmosAccountSpecifier)
        expect(selected).toEqual([
          'cosmosvaloper1qtxec3ggeuwnca9mmngw7vf6ctw54cppey02fs',
          'cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf',
          'cosmosvaloper1clpqr4nrk4khgkxj78fcwwh6dl3uw4epsluffn',
        ])
      })

      describe('validators and staking data are loaded', () => {
        beforeEach(() => {
          store.dispatch(
            stakingData.actions.upsertValidatorData({
              validators: mockValidatorData,
            }),
          )
        })

        it('returns empty array when all validators in stakingData are loaded / known', async () => {
          store.dispatch(
            stakingData.actions.upsertStakingData({
              accountSpecifier: cosmosAccountSpecifier,
              stakingData: mockStakingData,
            }),
          )

          const selected = selectNonloadedValidators(store.getState(), cosmosAccountSpecifier)
          expect(selected).toEqual([])
        })

        it('returns correct array when some validators in stakingData are not loaded / unknown', async () => {
          store.dispatch(
            stakingData.actions.upsertStakingData({
              accountSpecifier: cosmosAccountSpecifier,
              stakingData: mockStakingWithUnknownValidators,
            }),
          )

          const selected = selectNonloadedValidators(store.getState(), cosmosAccountSpecifier)
          expect(selected).toEqual([
            'cosmosvaloper1r9lxkpqre6j4487ut882xchgr7rdtx3x76gtdp',
            'cosmosvaloper1yvwqd5rdtuaw25mcqhz794dvgq9k9yeh8mjcdh',
          ])
        })
      })
    })

    describe('selectSingleValidator', () => {
      it('returns null on initial state', async () => {
        const selected = selectSingleValidator(
          store.getState(),
          cosmosAccountSpecifier,
          SHAPESHIFT_VALIDATOR_ADDRESS,
        )
        expect(selected).toBeNull()
      })

      describe('validators and staking data are loaded', () => {
        beforeEach(() => {
          store.dispatch(
            stakingData.actions.upsertValidatorData({
              validators: mockValidatorData,
            }),
          )
        })

        it('returns null when validator data is not present in state', async () => {
          store.dispatch(
            stakingData.actions.upsertStakingData({
              accountSpecifier: cosmosAccountSpecifier,
              stakingData: mockStakingData,
            }),
          )

          const selected = selectSingleValidator(
            store.getState(),
            cosmosAccountSpecifier,
            'cosmosvaloper1xxjvzwjpsf6ktuffkcpx29ut9qfrn0my8xdtqd',
          )
          expect(selected).toBeNull()
        })

        it('returns validator info when validator data is present in state', async () => {
          store.dispatch(
            stakingData.actions.upsertStakingData({
              accountSpecifier: cosmosAccountSpecifier,
              stakingData: mockStakingData,
            }),
          )

          const selected = selectSingleValidator(
            store.getState(),
            cosmosAccountSpecifier,
            SHAPESHIFT_VALIDATOR_ADDRESS,
          )
          expect(selected).toEqual({
            address: 'cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf',
            tokens: '111116',
            apr: '0.1496681491',
            commission: '0.100000000000000000',
            moniker: 'ShapeShift DAO',
          })
        })
      })
    })

    describe('selectRewardsAmountByAssetId', () => {
      it('returns empty string on initial state', async () => {
        const selected = selectRewardsAmountByAssetId(
          store.getState(),
          cosmosAccountSpecifier,
          SHAPESHIFT_VALIDATOR_ADDRESS,
          cosmosAssetId,
        )
        expect(selected).toEqual('')
      })

      describe('validators and staking data are loaded', () => {
        beforeEach(() => {
          store.dispatch(
            stakingData.actions.upsertValidatorData({
              validators: mockValidatorData,
            }),
          )
          store.dispatch(
            stakingData.actions.upsertStakingData({
              accountSpecifier: cosmosAccountSpecifier,
              stakingData: mockStakingData,
            }),
          )
        })

        it('returns empty string when there is no reward data', async () => {
          const selected = selectRewardsAmountByAssetId(
            store.getState(),
            cosmosAccountSpecifier,
            'cosmosvaloper1qtxec3ggeuwnca9mmngw7vf6ctw54cppey02fs',
            cosmosAssetId,
          )
          expect(selected).toEqual('')

          const selected2 = selectRewardsAmountByAssetId(
            store.getState(),
            cosmosAccountSpecifier,
            'cosmosvaloper1r9lxkpqre6j4487ut882xchgr7rdtx3x76gtdp',
            cosmosAssetId,
          )
          expect(selected2).toEqual('')
        })

        it('returns the reward amount when there is a reward', async () => {
          const selected = selectRewardsAmountByAssetId(
            store.getState(),
            cosmosAccountSpecifier,
            SHAPESHIFT_VALIDATOR_ADDRESS,
            cosmosAssetId,
          )
          expect(selected).toEqual('3.831752143667562385')
        })
      })
    })

    describe('selectTotalBondingsBalanceByAssetId', () => {
      it("returns '0' on initial state", async () => {
        const selected = selectTotalBondingsBalanceByAssetId(
          store.getState(),
          cosmosAccountSpecifier,
          SHAPESHIFT_VALIDATOR_ADDRESS,
          cosmosAssetId,
        )
        expect(selected).toEqual('0')
      })

      describe('validators and staking data are loaded', () => {
        beforeEach(() => {
          store.dispatch(
            stakingData.actions.upsertValidatorData({
              validators: mockValidatorData,
            }),
          )
          store.dispatch(
            stakingData.actions.upsertStakingData({
              accountSpecifier: cosmosAccountSpecifier,
              stakingData: mockStakingData,
            }),
          )
        })

        it('returns the reward amount when there is a reward', async () => {
          const selected = selectTotalBondingsBalanceByAssetId(
            store.getState(),
            cosmosAccountSpecifier,
            SHAPESHIFT_VALIDATOR_ADDRESS,
            cosmosAssetId,
          )
          expect(selected).toEqual('10115')

          const selected2 = selectTotalBondingsBalanceByAssetId(
            store.getState(),
            cosmosAccountSpecifier,
            'cosmosvaloper1qtxec3ggeuwnca9mmngw7vf6ctw54cppey02fs',
            cosmosAssetId,
          )
          expect(selected2).toEqual('4')
        })
      })
    })

    describe('selectTotalStakingDelegationCryptoByAccountSpecifier', () => {
      it('returns 0 on initial state', async () => {
        const selected = selectTotalStakingDelegationCryptoByAccountSpecifier(
          store.getState(),
          cosmosAccountSpecifier,
        )
        expect(selected).toEqual('0')
      })

      describe('validators and staking data are loaded', () => {
        beforeEach(() => {
          store.dispatch(
            stakingData.actions.upsertValidatorData({
              validators: mockValidatorData,
            }),
          )
          store.dispatch(
            stakingData.actions.upsertStakingData({
              accountSpecifier: cosmosAccountSpecifier,
              stakingData: mockStakingData,
            }),
          )
        })

        it('returns the total staking delegations amount for the account specifier', async () => {
          const selected = selectTotalStakingDelegationCryptoByAccountSpecifier(
            store.getState(),
            cosmosAccountSpecifier,
          )
          expect(selected).toEqual('15019')
        })
      })
    })
  })

  describe('selectStakingDataIsLoaded and selectValidatorIsLoaded', () => {
    it('returns false when nothing is loaded', async () => {
      const selected = selectStakingDataIsLoaded(store.getState())
      const selected2 = selectValidatorIsLoaded(store.getState())
      expect(selected).toBeFalsy()
      expect(selected2).toBeFalsy()
    })

    it('returns false when data is loading', async () => {
      store.dispatch(stakingData.actions.setStatus('loading'))
      const selected = selectStakingDataIsLoaded(store.getState())
      expect(selected).toBeFalsy()

      store.dispatch(stakingData.actions.setValidatorStatus('loading'))
      const selected2 = selectValidatorIsLoaded(store.getState())
      expect(selected2).toBeFalsy()
    })

    it('returns true when data is loaded', async () => {
      store.dispatch(stakingData.actions.setStatus('loaded'))
      const selected = selectStakingDataIsLoaded(store.getState())
      expect(selected).toBeTruthy()

      store.dispatch(stakingData.actions.setValidatorStatus('loaded'))
      const selected2 = selectValidatorIsLoaded(store.getState())
      expect(selected2).toBeTruthy()
    })
  })
})
