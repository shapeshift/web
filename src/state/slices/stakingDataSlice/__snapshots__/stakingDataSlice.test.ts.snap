// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stakingDataSlice reducers clear clears state back to initial state 1`] = `
Object {
  "delegations": Array [
    Object {
      "amount": "4",
      "assetId": "cosmos:cosmoshub-4/slip44:118",
      "validator": Object {
        "address": "cosmosvaloper1qtxec3ggeuwnca9mmngw7vf6ctw54cppey02fs",
        "apr": "0.1662979435",
        "commission": "0.000000000000000000",
        "moniker": "tokenpocket",
        "tokens": "111115",
      },
    },
    Object {
      "amount": "10015",
      "assetId": "cosmos:cosmoshub-4/slip44:118",
      "validator": Object {
        "address": "cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf",
        "apr": "0.1496681491",
        "commission": "0.100000000000000000",
        "moniker": "ShapeShift DAO",
        "tokens": "111116",
      },
    },
    Object {
      "amount": "5000",
      "assetId": "cosmos:cosmoshub-4/slip44:118",
      "validator": Object {
        "address": "cosmosvaloper1clpqr4nrk4khgkxj78fcwwh6dl3uw4epsluffn",
        "apr": "0.1514974265",
        "commission": "0.089000000000000000",
        "moniker": "Cosmostation",
        "tokens": "111117",
      },
    },
  ],
  "redelegations": Array [
    Object {
      "destinationValidator": Object {
        "address": "cosmosvaloper1qtxec3ggeuwnca9mmngw7vf6ctw54cppey02fs",
        "apr": "0.1662979435",
        "commission": "0.000000000000000000",
        "moniker": "tokenpocket",
        "tokens": "111115",
      },
      "entries": Array [
        Object {
          "amount": "4",
          "assetId": "cosmos:cosmoshub-4/slip44:118",
          "completionTime": 1650470407,
        },
      ],
      "sourceValidator": Object {
        "address": "cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf",
        "apr": "0.1496681491",
        "commission": "0.100000000000000000",
        "moniker": "ShapeShift DAO",
        "tokens": "111116",
      },
    },
  ],
  "rewards": Array [
    Object {
      "rewards": Array [],
      "validator": Object {
        "address": "cosmosvaloper1qtxec3ggeuwnca9mmngw7vf6ctw54cppey02fs",
        "apr": "0.1662979435",
        "commission": "0.000000000000000000",
        "moniker": "tokenpocket",
        "tokens": "111115",
      },
    },
    Object {
      "rewards": Array [
        Object {
          "amount": "3.831752143667562385",
          "assetId": "cosmos:cosmoshub-4/slip44:118",
        },
      ],
      "validator": Object {
        "address": "cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf",
        "apr": "0.1496681491",
        "commission": "0.100000000000000000",
        "moniker": "ShapeShift DAO",
        "tokens": "111116",
      },
    },
    Object {
      "rewards": Array [
        Object {
          "amount": "12.688084635379675000",
          "assetId": "cosmos:cosmoshub-4/slip44:118",
        },
      ],
      "validator": Object {
        "address": "cosmosvaloper1clpqr4nrk4khgkxj78fcwwh6dl3uw4epsluffn",
        "apr": "0.1514974265",
        "commission": "0.089000000000000000",
        "moniker": "Cosmostation",
        "tokens": "111117",
      },
    },
  ],
  "undelegations": Array [
    Object {
      "entries": Array [
        Object {
          "amount": "100",
          "assetId": "cosmos:cosmoshub-4/slip44:118",
          "completionTime": 1650472940,
        },
      ],
      "validator": Object {
        "address": "cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf",
        "apr": "0.1496681491",
        "commission": "0.100000000000000000",
        "moniker": "ShapeShift DAO",
        "tokens": "111116",
      },
    },
  ],
}
`;

exports[`stakingDataSlice reducers clear clears state back to initial state 2`] = `
Object {
  "cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf": Object {
    "address": "cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf",
    "apr": "0.1496681491",
    "commission": "0.100000000000000000",
    "moniker": "ShapeShift DAO",
    "tokens": "111116",
  },
  "cosmosvaloper1clpqr4nrk4khgkxj78fcwwh6dl3uw4epsluffn": Object {
    "address": "cosmosvaloper1clpqr4nrk4khgkxj78fcwwh6dl3uw4epsluffn",
    "apr": "0.1514974265",
    "commission": "0.089000000000000000",
    "moniker": "Cosmostation",
    "tokens": "111117",
  },
  "cosmosvaloper1fqzqejwkk898fcslw4z4eeqjzesynvrdfr5hte": Object {
    "address": "cosmosvaloper1fqzqejwkk898fcslw4z4eeqjzesynvrdfr5hte",
    "apr": "0.1513311286",
    "commission": "0.090000000000000000",
    "moniker": "commercio.network",
    "tokens": "11116",
  },
  "cosmosvaloper1gf3dm2mvqhymts6ksrstlyuu2m8pw6dhfp9md2": Object {
    "address": "cosmosvaloper1gf3dm2mvqhymts6ksrstlyuu2m8pw6dhfp9md2",
    "apr": "0.1496681491",
    "commission": "0.100000000000000000",
    "moniker": "Blockapsis",
    "tokens": "11115",
  },
  "cosmosvaloper1q6d3d089hg59x6gcx92uumx70s5y5wadklue8s": Object {
    "address": "cosmosvaloper1q6d3d089hg59x6gcx92uumx70s5y5wadklue8s",
    "apr": "0.1662979435",
    "commission": "0.000000000000000000",
    "moniker": "UbikCapital(0%Commission)",
    "tokens": "11114",
  },
  "cosmosvaloper1qdxmyqkvt8jsxpn5pp45a38ngs36mn2604cqk9": Object {
    "address": "cosmosvaloper1qdxmyqkvt8jsxpn5pp45a38ngs36mn2604cqk9",
    "apr": "0.1496681491",
    "commission": "0.100000000000000000",
    "moniker": "真本聪&IOSG",
    "tokens": "11113",
  },
  "cosmosvaloper1qtxec3ggeuwnca9mmngw7vf6ctw54cppey02fs": Object {
    "address": "cosmosvaloper1qtxec3ggeuwnca9mmngw7vf6ctw54cppey02fs",
    "apr": "0.1662979435",
    "commission": "0.000000000000000000",
    "moniker": "tokenpocket",
    "tokens": "111115",
  },
}
`;

exports[`stakingDataSlice reducers upsertStakingData updates or inserts empty staking data in state 1`] = `
Object {
  "delegations": Array [],
  "redelegations": Array [],
  "rewards": Array [],
  "undelegations": Array [],
}
`;

exports[`stakingDataSlice reducers upsertStakingData updates or inserts staking data in state 1`] = `
Object {
  "delegations": Array [
    Object {
      "amount": "4",
      "assetId": "cosmos:cosmoshub-4/slip44:118",
      "validator": Object {
        "address": "cosmosvaloper1qtxec3ggeuwnca9mmngw7vf6ctw54cppey02fs",
        "apr": "0.1662979435",
        "commission": "0.000000000000000000",
        "moniker": "tokenpocket",
        "tokens": "111115",
      },
    },
    Object {
      "amount": "10015",
      "assetId": "cosmos:cosmoshub-4/slip44:118",
      "validator": Object {
        "address": "cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf",
        "apr": "0.1496681491",
        "commission": "0.100000000000000000",
        "moniker": "ShapeShift DAO",
        "tokens": "111116",
      },
    },
    Object {
      "amount": "5000",
      "assetId": "cosmos:cosmoshub-4/slip44:118",
      "validator": Object {
        "address": "cosmosvaloper1clpqr4nrk4khgkxj78fcwwh6dl3uw4epsluffn",
        "apr": "0.1514974265",
        "commission": "0.089000000000000000",
        "moniker": "Cosmostation",
        "tokens": "111117",
      },
    },
  ],
  "redelegations": Array [
    Object {
      "destinationValidator": Object {
        "address": "cosmosvaloper1qtxec3ggeuwnca9mmngw7vf6ctw54cppey02fs",
        "apr": "0.1662979435",
        "commission": "0.000000000000000000",
        "moniker": "tokenpocket",
        "tokens": "111115",
      },
      "entries": Array [
        Object {
          "amount": "4",
          "assetId": "cosmos:cosmoshub-4/slip44:118",
          "completionTime": 1650470407,
        },
      ],
      "sourceValidator": Object {
        "address": "cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf",
        "apr": "0.1496681491",
        "commission": "0.100000000000000000",
        "moniker": "ShapeShift DAO",
        "tokens": "111116",
      },
    },
  ],
  "rewards": Array [
    Object {
      "rewards": Array [],
      "validator": Object {
        "address": "cosmosvaloper1qtxec3ggeuwnca9mmngw7vf6ctw54cppey02fs",
        "apr": "0.1662979435",
        "commission": "0.000000000000000000",
        "moniker": "tokenpocket",
        "tokens": "111115",
      },
    },
    Object {
      "rewards": Array [
        Object {
          "amount": "3.831752143667562385",
          "assetId": "cosmos:cosmoshub-4/slip44:118",
        },
      ],
      "validator": Object {
        "address": "cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf",
        "apr": "0.1496681491",
        "commission": "0.100000000000000000",
        "moniker": "ShapeShift DAO",
        "tokens": "111116",
      },
    },
    Object {
      "rewards": Array [
        Object {
          "amount": "12.688084635379675000",
          "assetId": "cosmos:cosmoshub-4/slip44:118",
        },
      ],
      "validator": Object {
        "address": "cosmosvaloper1clpqr4nrk4khgkxj78fcwwh6dl3uw4epsluffn",
        "apr": "0.1514974265",
        "commission": "0.089000000000000000",
        "moniker": "Cosmostation",
        "tokens": "111117",
      },
    },
  ],
  "undelegations": Array [
    Object {
      "entries": Array [
        Object {
          "amount": "100",
          "assetId": "cosmos:cosmoshub-4/slip44:118",
          "completionTime": 1650472940,
        },
      ],
      "validator": Object {
        "address": "cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf",
        "apr": "0.1496681491",
        "commission": "0.100000000000000000",
        "moniker": "ShapeShift DAO",
        "tokens": "111116",
      },
    },
  ],
}
`;

exports[`stakingDataSlice reducers upsertValidatorData updates or inserts validator data in state 1`] = `
Object {
  "cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf": Object {
    "address": "cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf",
    "apr": "0.1496681491",
    "commission": "0.100000000000000000",
    "moniker": "ShapeShift DAO",
    "tokens": "111116",
  },
  "cosmosvaloper1clpqr4nrk4khgkxj78fcwwh6dl3uw4epsluffn": Object {
    "address": "cosmosvaloper1clpqr4nrk4khgkxj78fcwwh6dl3uw4epsluffn",
    "apr": "0.1514974265",
    "commission": "0.089000000000000000",
    "moniker": "Cosmostation",
    "tokens": "111117",
  },
  "cosmosvaloper1fqzqejwkk898fcslw4z4eeqjzesynvrdfr5hte": Object {
    "address": "cosmosvaloper1fqzqejwkk898fcslw4z4eeqjzesynvrdfr5hte",
    "apr": "0.1513311286",
    "commission": "0.090000000000000000",
    "moniker": "commercio.network",
    "tokens": "11116",
  },
  "cosmosvaloper1gf3dm2mvqhymts6ksrstlyuu2m8pw6dhfp9md2": Object {
    "address": "cosmosvaloper1gf3dm2mvqhymts6ksrstlyuu2m8pw6dhfp9md2",
    "apr": "0.1496681491",
    "commission": "0.100000000000000000",
    "moniker": "Blockapsis",
    "tokens": "11115",
  },
  "cosmosvaloper1q6d3d089hg59x6gcx92uumx70s5y5wadklue8s": Object {
    "address": "cosmosvaloper1q6d3d089hg59x6gcx92uumx70s5y5wadklue8s",
    "apr": "0.1662979435",
    "commission": "0.000000000000000000",
    "moniker": "UbikCapital(0%Commission)",
    "tokens": "11114",
  },
  "cosmosvaloper1qdxmyqkvt8jsxpn5pp45a38ngs36mn2604cqk9": Object {
    "address": "cosmosvaloper1qdxmyqkvt8jsxpn5pp45a38ngs36mn2604cqk9",
    "apr": "0.1496681491",
    "commission": "0.100000000000000000",
    "moniker": "真本聪&IOSG",
    "tokens": "11113",
  },
  "cosmosvaloper1qtxec3ggeuwnca9mmngw7vf6ctw54cppey02fs": Object {
    "address": "cosmosvaloper1qtxec3ggeuwnca9mmngw7vf6ctw54cppey02fs",
    "apr": "0.1662979435",
    "commission": "0.000000000000000000",
    "moniker": "tokenpocket",
    "tokens": "111115",
  },
}
`;

exports[`stakingDataSlice selectors selectActiveStakingOpportunityDataByAssetId validators and staking data are loaded returns correct array with delegations, undelegations, redelegations and rewards 1`] = `
Array [
  Object {
    "address": "cosmosvaloper1qtxec3ggeuwnca9mmngw7vf6ctw54cppey02fs",
    "apr": "0.1662979435",
    "cryptoAmount": "4",
    "moniker": "tokenpocket",
    "rewards": "0",
    "tokens": "111115",
  },
  Object {
    "address": "cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf",
    "apr": "0.1496681491",
    "cryptoAmount": "10115",
    "moniker": "ShapeShift DAO",
    "rewards": "3.831752143667562385",
    "tokens": "111116",
  },
  Object {
    "address": "cosmosvaloper1clpqr4nrk4khgkxj78fcwwh6dl3uw4epsluffn",
    "apr": "0.1514974265",
    "cryptoAmount": "5000",
    "moniker": "Cosmostation",
    "rewards": "12.688084635379675",
    "tokens": "111117",
  },
]
`;

exports[`stakingDataSlice selectors selectActiveStakingOpportunityDataByAssetId validators and staking data are loaded returns correct array with only rewards 1`] = `
Array [
  Object {
    "address": "cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf",
    "apr": "0.1496681491",
    "cryptoAmount": "0",
    "moniker": "ShapeShift DAO",
    "rewards": "4.831752143667562385",
    "tokens": "111116",
  },
  Object {
    "address": "cosmosvaloper1clpqr4nrk4khgkxj78fcwwh6dl3uw4epsluffn",
    "apr": "0.1514974265",
    "cryptoAmount": "0",
    "moniker": "Cosmostation",
    "rewards": "12.688084635379675",
    "tokens": "111117",
  },
]
`;

exports[`stakingDataSlice selectors selectActiveStakingOpportunityDataByAssetId validators and staking data are loaded returns correct array with only undelegations 1`] = `
Array [
  Object {
    "address": "cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf",
    "apr": "0.1496681491",
    "cryptoAmount": "350",
    "moniker": "ShapeShift DAO",
    "rewards": "0",
    "tokens": "111116",
  },
]
`;
